btfgen.c

```diff
@@ -280,56 +348,102 @@ static struct btf_reloc_type *btf_reloc_put_type(struct btf *btf,
-		printf("unsupported relocation: %d\n", reloc_type->id);
+		printf("unsupported kind: %s (%d)\n", btf_kind_str(btf_kind(reloc_type->type)), reloc_type->id);
 		return ERR_PTR(-EINVAL);
 	}
 
 	return reloc_type;
 }
 
-/* Return pointer to btf_reloc_type by id */
-static struct btf_reloc_type *btf_reloc_get_type(struct btf_reloc_info *info, int id) {
-	struct btf_reloc_type *type = NULL;
+/*
+ * Same as btf_reloc_put_type, but adding all fields, from given complex type, recursively
+ */
+static struct btf_reloc_type *btf_reloc_put_type_all(struct btf *btf,
+						     struct btf_reloc_info *info,
+						     struct btf_type *btf_type,
+						     unsigned int id) {
 
-	if (!hashmap__find(info->types, uint_as_hash_key(id), (void **)&type))
-		return ERR_PTR(-ENOENT);
+	struct btf_reloc_type *reloc_type, *tmp;
+	struct btf_array *array;
+	struct btf_member *m;
+	unsigned int child_id;
+	int err, i, n;
 
-	return type;
-}
+	if (!hashmap__find(info->types, uint_as_hash_key(id), (void**) &reloc_type)) {
+		reloc_type = calloc(1, sizeof(*reloc_type));
+		if (!reloc_type)
+			return ERR_PTR(-ENOMEM);
 
-static int bpf_reloc_type_add_member(struct btf_reloc_info *info,
-				     struct btf_reloc_type *reloc_type,
-				     struct btf_member *btf_member, int idx) {
-	int err;
-	struct btf_reloc_member *reloc_member;
+		reloc_type->type = btf_type;
+		reloc_type->id = id;
 
-	/* create new members hashmap for this relocation type if needed */
-	if (reloc_type->members == NULL) {
-		struct hashmap *tmp = hashmap__new(bpf_reloc_info_hash_fn,
-						   bpf_reloc_info_equal_fn,
-						   NULL);
+		err = hashmap__add(info->types, uint_as_hash_key(reloc_type->id), reloc_type);
+		if (err)
+			return ERR_PTR(err);
+	}
+
+	switch (btf_kind(reloc_type->type)) {
+	case BTF_KIND_INT:
+	case BTF_KIND_FLOAT:
+		/* not a complex type, already solved */
+		break;
+	case BTF_KIND_STRUCT:
+	case BTF_KIND_UNION:
+		n = btf_vlen(btf_type);
+
+		/* add all member types */
+		m = btf_members(btf_type);
+		for (i = 0; i < n; i++, m++) {
+			btf_type = (struct btf_type *) btf__type_by_id(btf, m->type);
+			if (!btf_type)
+				return ERR_PTR(-EINVAL);
+
+			tmp = btf_reloc_put_type_all(btf, info, btf_type, m->type);
+			if (IS_ERR(tmp))
+				return tmp;
+		}
+
+		/* add all members */
+		m = btf_members(btf_type);
+		for (i = 0; i < n; i++, m++) {
+			err = bpf_reloc_type_add_member(info, reloc_type, m, i);
+			if (err)
+				return ERR_PTR(err);
+		}
+		break;
+	case BTF_KIND_PTR:
+	case BTF_KIND_CONST:
+	case BTF_KIND_VOLATILE:
+	case BTF_KIND_TYPEDEF:
+		/* modifier types */
+		child_id = btf_type->type;
+		btf_type = (struct btf_type *) btf__type_by_id(btf, child_id);
+		if (!btf_type)
+			return ERR_PTR(-EINVAL);
+
+		tmp = btf_reloc_put_type_all(btf, info, btf_type, child_id);
 		if (IS_ERR(tmp))
-			return PTR_ERR(tmp);
+			return tmp;
+		break;
+	case BTF_KIND_ARRAY:
+		array = btf_array(btf_type);
+		btf_type = (struct btf_type *) btf__type_by_id(btf, array->type);
```

Comment:
I have not tested array with type based relocations (yet ?). And, yes, very nice catch. I'm not adding the index_type BTF type and, if it doesn't exist, it would generate an issue. Will fix that.